semaphore/semaphore.go

```diff
@@ -0,0 +1,71 @@
+// Package sempaphore provides a simple implementation of a semaphore (hence the name, duh).
+// Semaphores may be usefull to enforce a upper bound on how much
+// concurrent units executes some algorithm that is very heavy on memory and/or CPU.
+//
+// One example is when you have an http server that can accept hundreds
+// of concurrent requests but you have some critical section of your code that
+// cant handle that much concurrency at once (but you don't want to serialize either).
+//
+// There are other ways to handle this too using channels, which approach to use will
+// be your judgement call. If you choose the semaphore here you got one.
+package semaphore
+
+import (
+	"fmt"
+	"context"
+)
+
+// S is a semaphore instance. Always use the New function
+// to create semaphores. Using a unproperly initialized
+// semaphore may cause mayhem on your code.
+type S chan struct{}
+
+// Release is used to release a previous call to S.Acquire
+type Release func()
+
+// New creates a new semaphore with the given size.
+// Passing 0 as size is a moronic programming mistake and will
+// result in a panic due to its moronicness.
+func New(size uint) S {
+	if size == 0 {
+		panic("semaphore.New:cant create a semaphore with size 0")
+	}
+	return S(make(chan struct{}, size))
+}
+
+// Acquire will acquire the semaphore, if the semaphore is
+// already full Acquire will wait until some other goroutine
+// releases the semaphore or until the given context timeouts.
+//
+// If the context timeouts or is cancelled a non nil error is returned.
+// On success a release function is returned, this function can be used
+// only to release the correspondent call to Acquire, calling the
+// returned Release function twice will lock your goroutine forever since
+// this is a very stupid thing to do (better than releasing the semaphore twice
```

Comment:
I think only saying that calling release function twice is a programmer mistake that locks the goroutine suffices.

Reply:
In the end I implemented a panic instead of locking, but I'm not sure if it is the best way to Go. I dislike panicing but in this case is an extremelly severe programming error (like dividing by zero)... I think

Reply:
I was referring to the 'is a very stupid thing to do' part of the phrase, because it poses so much personal feelings on the documentation making it non professional. I have made this in the past and now I'm trying to improve this.
Feel free to ignore also, but nowadays for me, mature docs is as important as mature code.

Reply:
I don't think that mature goes hand in hand with being formal. Microsoft documentation is extremelly formal and has no jokes at all and the code sux ass and nothing works properly =P, so I always enjoyed some jokes on stuff to break the ice, never see it as a sign of unprofessional or imature work (lack of testing would do it...or how people handle bug reports...like my problems with Azure ¬¬).

But I can see your point...but on this project I was not aiming to being "professional" =P, just making things that I need and that works well...with tests...and stuff =D

Anyway, I will remove it, don't think the jokes add anything anyway (perhaps to someone else like me who enjoy this kind of crap =) ).